MARXAN-1494-exportId-public-project

rubvalave opened the pull request:
-Adds exportId column to public project.
-Creates export everytime a project is sent to publish.

hotzevzl commented on api/apps/api/src/modules/published-project/published-project.service.ts:
```diff
@@ -64,6 +68,25 @@ export class PublishedProjectService {
+        return scenario?.id;
+      });
+
+    const exportResult = await this.projectService.requestExport(
```
this is good - from the point of view of a user, by now we only know if the request of an export was queued and if we have an `exportId`: if the export will take "a long time" though, initial attempts to clone the exported (or rather, being-exported) project will fail (I think)... do we need to handle this? (not here and not in this PR, but at least in the endpoint to request a clone based on a given `exportId`)

rubvalave replied:
Maybe this is what @aciddaute was asking me about some minutes ago? Maybe we need a small meeting to detail the steps because I'm assuming the export works in a specific way that is not exactly what should be covered when publishing?

aciddaute replied:
`/api/v1/projects/published-projects/:exportId/clone` endpoint checks wether the export has finished or not

rubvalave commented on api/apps/api/src/modules/published-project/dto/publish-project.dto.ts:
```diff
@@ -41,4 +41,11 @@ export class CreatePublishProjectDto extends PublishProjectDto {
+  @IsOptional()
+  @IsArray()
+  @IsString({ each: true })
+  @IsUUID('all', { each: true })
```
Copy-paste went wrong (though I think this is as it is here in another dto of the system). I'll fix it.

hotzevzl replied:
yep, the other uses of `all` UUID variants are overly permissive too (not much relevant in practice, but we should also change those to v4 only... not in this PR!

rubvalave commented on api/apps/api/src/modules/published-project/published-project.service.ts:
```diff
@@ -64,6 +68,25 @@ export class PublishedProjectService {
       return left(alreadyPublished);
     }
 
+    const scenarioQuery = `SELECT id from scenarios where scenarios.project_id = $1`;
+    const scenarioIds = await this.projectRepository.query(scenarioQuery, [id]);
```
I had some circular dependencies errors yesterday and I realized with the projectRepo I was able to overcome those and end up doing the query I wanted. But you are right and I should be ashamed of this.

On the other hand, the scenarioIds bit is that if it is not set up by FE, the default should be taking all of them from a project. Maybe I should have clarified the names as well a little bit. I'll handle this! 👍

rubvalave commented on api/apps/api/src/migrations/api/1650984916463-AddsExportIdToPublicProject.ts:
```diff
@@ -0,0 +1,18 @@
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(`
+          ALTER TABLE published_projects
+            ADD COLUMN export_id varchar;
```
I don't think so, the `published-project` entity in the end is a copy of the actual `project`. I made this question as well regarding the `scenarioId` used for the preview map generation a while ago, and in the end linking this kind of copycat entity wasn't needed.

In fact, when you clone the public-project you are actually copying a `project` so I think it will be a little bit overkill if we link those two tables, but maybe @hotzevzl or @aagm have strong opinions about this here.

rubvalave commented:
@aciddaute and @hotzevzl I addressed most of the comments you made on the latest changes. I would like if you can review them and see if now it aligns better with your feedback. Thank you!
